### PR TITLE
XP-PEN G430 Driver: init at 20190820

### PR DESCRIPTION
**Pull request by IvarWithoutBones**
###### Motivation for this change
This driver is an useful application for people who own the XP-PEN G430 drawing tablet, as the tablet does not work without this.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

I'm very sorry if i messed something up, fairly new to this.

### CODE REVIEW

File: pkgs/misc/drivers/xp-pen-g430/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      libXtst
+      qt5.qtbase
+      libglvnd
+      stdenv.cc.cc.lib
+      ];
+  in ''
+    patchelf \
+      --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" \
+      --set-rpath "${libPath}" \
+      $out/bin/Pentablet_Driver
```
Comment (veprbl) [inline pkgs/misc/drivers/xp-pen-g430/default.nix]: Did you try using `autoPatchelfHook`?

Comment (Yvar192) [reply pkgs/misc/drivers/xp-pen-g430/default.nix]: I have not, as its a tad unstable in my experience.

Comment (veprbl) [reply pkgs/misc/drivers/xp-pen-g430/default.nix]: Could you try?

File: pkgs/misc/drivers/xp-pen-g430/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = https://www.xp-pen.com/download-46.html;
+    description = "Driver for the XP-PEN G430 drawing tablet";
+    license = licenses.unfree;
+    platforms = platforms.linux;
+    maintainers = [ Yvar192 ];
```
Comment (veprbl) [inline pkgs/misc/drivers/xp-pen-g430/default.nix]:
```suggestion
    maintainers = with maintainers; [ Yvar192 ];
```
Need to also add yourself to the maintainers list.

File: pkgs/misc/drivers/xp-pen-g430/default.nix (hunk 1 above)
Comment (Yvar192) [reply pkgs/misc/drivers/xp-pen-g430/default.nix]: Should now all be working with `autoPatchelfHook`.

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -25798,6 +25798,8 @@ in
 
   sentencepiece = callPackage ../development/libraries/sentencepiece {};
 
+  xp-pen-g430 = callPackage ../misc/drivers/xp-pen-g430 {};
```
Comment (veprbl) [inline pkgs/top-level/all-packages.nix]: Should use `libsForQt5.callPackage`
https://nixos.org/nixpkgs/manual/#sec-language-qt

File: pkgs/misc/drivers/xp-pen-g430/default.nix [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    stdenv.cc.cc.lib
+  ];
+
+  unpackPhase = ''
+    tar -xzf Linux_Pentablet_V1.3.0.0.tar.gz
+  '';
+
+  installPhase = ''
+    mkdir -p $out/bin
+    cp Linux_Pentablet_V1.3.0.0/Pentablet_Driver Linux_Pentablet_V1.3.0.0/config.xml $out/bin
```
Comment (Profpatsch) [inline pkgs/misc/drivers/xp-pen-g430/default.nix]: huh, shouldn’t this be in `share/doc` or something?

Comment (Yvar192) [reply pkgs/misc/drivers/xp-pen-g430/default.nix]: the xml is a config file required to be next to the executable to run

File: pkgs/misc/drivers/xp-pen-g430/default.nix [hunk 5]
```diff
@@ -0,0 +1,41 @@
+{ stdenv, fetchzip, tar, autoPatchelfHook, libusb1, libX11, libXtst, qt5, libglvnd }:
```
Comment (veprbl) [inline pkgs/misc/drivers/xp-pen-g430/default.nix]:
```suggestion
{ stdenv, fetchzip, tar, autoPatchelfHook, libusb1, libX11, libXtst, qtbase, libglvnd }:
```

File: pkgs/misc/drivers/xp-pen-g430/default.nix [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  nativeBuildInputs = [
+    tar
+    autoPatchelfHook
+  ];
+
+  buildInputs = [
+    libusb1
+    libX11
+    libXtst
+    qt5.qtbase
```
Comment (veprbl) [inline pkgs/misc/drivers/xp-pen-g430/default.nix]:
```suggestion
    qtbase
```

File: pkgs/misc/drivers/xp-pen-g430/default.nix (hunk 5 above)
Comment (veprbl) [reply pkgs/misc/drivers/xp-pen-g430/default.nix]:
```
anonymous function at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/eval-0-gleber.ewr1.nix.ci/pkgs/misc/drivers/xp-pen-g430/default.nix:1:1 called without required argument 'tar', at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/eval-0-gleber.ewr1.nix.ci/lib/customisation.nix:69:16
```
How does this work for you?

File: pkgs/misc/drivers/xp-pen-g430/default.nix [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    libglvnd
+    stdenv.cc.cc.lib
+  ];
+
+  unpackPhase = ''
+    tar -xzf Linux_Pentablet_V1.3.0.0.tar.gz
```
Comment (veprbl) [inline pkgs/misc/drivers/xp-pen-g430/default.nix]: Probably can just do
```
src = fetchzip {
  url = "https://download01.xp-pen.com/file/2019/08/Linux%20Beta%20Driver(${version}).zip";
  sha256 = "091kfqxxj90pdmwncgfl8ldi70pdhwryh3cls30654983m8cgnby";
} + ./Linux_Pentablet_V1.3.0.0.tar.gz;
```
and drop the custom unpackPhase

Comment (IvarWithoutBones) [reply pkgs/misc/drivers/xp-pen-g430/default.nix]: Was not aware this was possible, will do :+1:

**Comment by IvarWithoutBones** [comment]
When trying to build locally with the change from `qt5.qtebase` > `qtbase` i get the following error:
`error: anonymous function at /home/ivar/misc/nixpkgs/pkgs/misc/drivers/xp-pen-g430/default.nix:1:1 called without required argument 'qtbase', at /nix/store/lxid8qkmifw0bhyxa14qanvmkv5lkd5d-nixpkgs-20.09pre217039.13e7a3e1127/nixpkgs/lib/customisation.nix:69:16`
Not sure how to fix

File: pkgs/misc/drivers/xp-pen-g430/default.nix [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    stdenv.cc.cc.lib
+  ];
+
+  installPhase = ''
+    mkdir -p $out/bin
+    cp Linux_Pentablet_V1.3.0.0/Pentablet_Driver Linux_Pentablet_V1.3.0.0/config.xml $out/bin
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = https://www.xp-pen.com/download-46.html;
```
Comment (veprbl) [inline pkgs/misc/drivers/xp-pen-g430/default.nix]:
```suggestion
    homepage = "https://www.xp-pen.com/download-46.html";
```

File: pkgs/misc/drivers/xp-pen-g430/default.nix [hunk 9]
```diff
@@ -0,0 +1,36 @@
+{ stdenv, fetchzip, autoPatchelfHook, libusb1, libX11, libXtst, qtbase, libglvnd }:
+stdenv.mkDerivation rec {
+  name = "xp-pen-G430";
+  version = "20190820";
+
+  src = fetchzip {
+    url = "https://download01.xp-pen.com/file/2019/08/Linux%20Beta%20Driver(${version}).zip";
+    sha256 = "091kfqxxj90pdmwncgfl8ldi70pdhwryh3cls30654983m8cgnby";
+  } + ./Linux_Pentablet_V1.3.0.0.tar.gz;
```
Comment (veprbl) [inline pkgs/misc/drivers/xp-pen-g430/default.nix]:
```suggestion
  } + /Linux_Pentablet_V1.3.0.0.tar.gz;
```

File: pkgs/misc/drivers/xp-pen-g430/default.nix [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    libX11
+    libXtst
+    qtbase
+    libglvnd
+    stdenv.cc.cc.lib
+  ];
+
+  installPhase = ''
+    mkdir -p $out/bin
+    cp Linux_Pentablet_V1.3.0.0/Pentablet_Driver Linux_Pentablet_V1.3.0.0/config.xml $out/bin
```
Comment (veprbl) [inline pkgs/misc/drivers/xp-pen-g430/default.nix]:
```suggestion
    cp {Pentablet_Driver,config.xml} "$out"/bin
```

**Review by veprbl — COMMENTED** [review]
Please test your PR

File: pkgs/misc/drivers/xp-pen-g430/default.nix [hunk 11]
```diff
@@ -0,0 +1,37 @@
+{ stdenv, fetchzip, autoPatchelfHook, libusb1, libX11, libXtst, qtbase, wrapQtAppsHook, libglvnd }:
+stdenv.mkDerivation rec {
+  name = "xp-pen-G430";
```
Comment (veprbl) [inline pkgs/misc/drivers/xp-pen-g430/default.nix]:
```suggestion
  pname = "xp-pen-G430";
```

**Review by veprbl — COMMENTED** [review]
Please squash all commits and rebase on top of master

**Comment by IvarWithoutBones** [comment]
Rebased, any chance you could just use the squash&merge feature from github? Im not that familiar with git, sorry :frowning_face:

**Comment by Profpatsch** [comment]
Let me take a look

**Comment by Profpatsch** [comment]
I’m sorry, this PR is too broken to fix up. Could you manually take your current state and force-push it as one commit to this branch? Or open a new PR if you can’t figure out how to force-push (I admit, git can be a bit confusing).

Github has nice guides that give an introduction on how to use git with it’s interface, maybe you will want to read those.

As a general rule of thumb: don’t merge master back into your pull request branch, it makes changing things after the fact very hard.

**Comment by IvarWithoutBones** [comment]
Apologies for the trouble, will open up a new PR.

**Pull request closed without merging**